src/develop/pixelpipe_hb.c

```diff
@@ -334,42 +334,45 @@ void dt_dev_pixelpipe_synch(dt_dev_pixelpipe_t *pipe, dt_develop_t *dev, GList *
   dt_dev_history_item_t *hist = (dt_dev_history_item_t *)history->data;
   // find piece in nodes list
   dt_dev_pixelpipe_iop_t *piece = NULL;
-  gboolean hint = FALSE;
+
+  const dt_image_t *img = &pipe->image;
+  const int32_t imgid = img->id;
+  const gboolean rawprep_img = dt_image_is_rawprepare_supported(img);
 
   for(GList *nodes = pipe->nodes; nodes; nodes = g_list_next(nodes))
   {
     piece = (dt_dev_pixelpipe_iop_t *)nodes->data;
-    const dt_image_t *img = &piece->pipe->image;
-    const int imgid = img->id;
 
     if(piece->module == hist->module)
     {
-      gboolean active = hist->enabled;
+      const gboolean active = hist->enabled;
+      piece->enabled = active;
 
-      // demosaic must be OFF for non-raws and ON for raws
-      if(strcmp(piece->module->op, "demosaic") == 0)
+      // Styles, presets or history copy&paste might set history items not appropriate for the image.
+      // Fixing that seemed to be almost impossible after long discussions but at least we can test,
+      // correct and add a problem hint here.
+      if((strcmp(piece->module->op, "demosaic") == 0) || (strcmp(piece->module->op, "rawprepare") == 0))
       {
-        if(dt_image_is_raw(img) && !active)
-        {
-          hint = TRUE;
-          active = TRUE;
-          fprintf(stderr,"[dt_dev_pixelpipe_synch] found disabled demosaic in history for raw `%s`, id: %i\n",
-            img->filename, imgid);
-        }
-        else if(!dt_image_is_raw(img) && active)
-        {
-          hint = TRUE;
-          active = FALSE;
-          fprintf(stderr,"[dt_dev_pixelpipe_synch] found enabled demosaic in history for non-raw `%s`, id: %i\n",
-            img->filename, imgid);
-        }
+        if(rawprep_img && !active)
+          piece->enabled = TRUE;
+        else if(!rawprep_img && active)
+          piece->enabled = FALSE;
+      }
+      else if((strcmp(piece->module->op, "rawdenoise") == 0) ||
+              (strcmp(piece->module->op, "hotpixels") == 0) ||
+              (strcmp(piece->module->op, "cacorrect") == 0))
+      {
+        if(!rawprep_img && active) piece->enabled = FALSE;
+      }
+
+      if(piece->enabled != hist->enabled)
+      {
+        dt_iop_set_module_trouble_message(piece->module, _("problem"), _("usage of the module is not correct for this image.\nlikely introduced by applying a preset, style or history copy&paste"), NULL);
```

Comment:
Instead of just `problem` in red in the iop module:

![image](https://user-images.githubusercontent.com/467069/113822972-61f43580-977e-11eb-93a9-ff13865b3741.png)

I would write "disabled as issue detected" or "enabled as required" depending on `piece->enabled` value.

What do you think?

Reply:
really good